### PR TITLE
Harden DiscoverModelMetadata: fix hasPooledOutput, scope fallback, extract helper

### LINKED ISSUE #28: Harden DiscoverModelMetadata: fix hasPooledOutput detection, scope MaxTokenLength fallback, extract helper
## Summary

`DiscoverModelMetadata()` in `OnnxTextModelScorerEstimator.cs` has several robustness issues identified during PR #26 code review. None cause failures with currently known models, but they represent latent bugs that could surface with non-standard ONNX exports.

Related: #25 (the fix that surfaced these), PR #26 (merged fix)

---

## Issue 1: `hasPooledOutput` detection rejects dynamic batch dimensions (pre-existing bug)

**Lines affected:** ~187, ~200, ~210

**Current logic:**
```csharp
hasPooledOutput = !dims.Contains(-1) && dims.Length == 2;
```

**The bug:** A pooled output like `sentence_embedding` with shape `[-1, 384]` (dynamic batch, static hidden) gets `hasPooledOutput = false` because `dims.Contains(-1)` is true (the batch dim is -1). This would cause the embedding facade to attempt pooling on an already-pooled vector.

**Why it doesn't fail today:** The auto-discovery branch (line ~223) sets `hasPooledOutput = true` directly when it finds `sentence_embedding`/`pooler_output`. The buggy `!dims.Contains(-1)` check only exists in the `OutputTensorName` and `PreferredOutputNames` branches, which aren't typically used for pooled embedding outputs. But if someone sets `PreferredOutputNames = ["sentence_embedding"]`, it would incorrectly report `hasPooledOutput = false`.

**Fix:**
```csharp
// Option A: Check only that the last dim is static and rank is 2
hasPooledOutput = dims.Length == 2 && dims.Last() > 0;

// Option B: Check by known tensor names
hasPooledOutput = dims.Length == 2 && 
    (outputName == "sentence_embedding" || outputName == "pooler_output" || !dims.Contains(-1));
```

Option A is simpler and more correct  the batch dimension being dynamic is completely normal and doesn't affect whether the output is pre-pooled.

**Files to modify:** `src/MLNet.TextInference.Onnx/OnnxTextModelScorerEstimator.cs`
- Line ~187: `OutputTensorName` branch
- Line ~200: `PreferredOutputNames` branch (preferred found)
- Line ~210: `PreferredOutputNames` branch (preferred not found)

---

## Issue 2: `MaxTokenLength` fallback is semantically wrong for pooled and rank-3 outputs

**Lines affected:** ~225-226 (pooled), ~236-237 (rank-3 unpooled)

**The problem:** PR #26 added `if (hiddenDim <= 0) hiddenDim = _options.MaxTokenLength;` to all 5 branches. This is correct for **rank-2 logit outputs** (QA models) where the last dim = sequence_length. But for:

- **Pooled outputs** (`sentence_embedding`): last dim = embedding dimension (e.g., 384), NOT sequence length
- **Rank-3 unpooled outputs** (`last_hidden_state`): last dim = hidden dimension (e.g., 384), NOT sequence length

Falling back to `MaxTokenLength` (128) for these would silently produce wrong output sizes.

**Why it doesn't fail today:** No known ONNX model has a dynamic hidden dimension for `sentence_embedding` or `last_hidden_state`. The hidden dim is always a static architecture parameter (384, 768, etc.). So the fallback never triggers.

**Fix:** Replace the `MaxTokenLength` fallback with a throw in the pooled and rank-3 branches:

```csharp
// Pooled branch (line ~225):
hiddenDim = (int)outputMeta[pooledName].Dimensions.Last();
if (hiddenDim <= 0)
    throw new InvalidOperationException(
        $"Pooled ONNX output '{pooledName}' has a dynamic hidden dimension, which is not supported. " +
        $"Set OutputTensorName explicitly or use a model with a static embedding dimension.");

// Rank-3 unpooled branch (line ~236):
hiddenDim = (int)outputMeta[outputName].Dimensions.Last();
if (hiddenDim <= 0)
    throw new InvalidOperationException(
        $"ONNX output '{outputName}' has a dynamic hidden dimension, which is not supported for auto-discovery. " +
        $"Set OutputTensorName explicitly or use a model with a static hidden dimension.");
```

Keep the `MaxTokenLength` fallback ONLY in branches where rank == 2 and the output is NOT pooled (i.e., QA logit outputs).

**Files to modify:** `src/MLNet.TextInference.Onnx/OnnxTextModelScorerEstimator.cs`
- Line ~225-226: Pooled auto-discovery branch  throw instead of fallback
- Line ~236-237: Rank-3 unpooled auto-discovery branch  throw instead of fallback

---

## Issue 3: Extract helper to reduce duplication (refactoring)

**Current state:** The pattern `hiddenDim = (int)dims.Last(); if (hiddenDim <= 0) hiddenDim = _options.MaxTokenLength;` is repeated in 5 branches.

**Suggested refactoring:**
```csharp
private int ResolveLastDimension(long[] dims, string tensorName, bool allowDynamicFallback)
{
    int lastDim = (int)dims.Last();
    if (lastDim > 0)
        return lastDim;
    
    if (allowDynamicFallback)
        return _options.MaxTokenLength;
    
    throw new InvalidOperationException(
        $"ONNX output '{tensorName}' has a dynamic last dimension, which is not supported for this output type.");
}
```

Then each branch calls:
```csharp
// QA logit outputs (rank 2, not pooled): allow fallback
hiddenDim = ResolveLastDimension(dims, outputName, allowDynamicFallback: true);

// Pooled/rank-3 outputs: fail fast
hiddenDim = ResolveLastDimension(dims, outputName, allowDynamicFallback: false);
```

This combines the fix for Issue 2 with the deduplication.

**Files to modify:** `src/MLNet.TextInference.Onnx/OnnxTextModelScorerEstimator.cs`

---

### Test Issue 1 (hasPooledOutput)
1. Create a mock scenario where `PreferredOutputNames = ["sentence_embedding"]` points to a tensor with shape `[-1, 384]`
2. Verify `hasPooledOutput` is `true` (currently would be `false`)

### Test Issue 2 (fallback semantics)
1. Create a mock scenario with a pooled output tensor where last dim is -1
2. Verify it throws `InvalidOperationException` instead of silently using `MaxTokenLength`
3. Verify QA logit outputs (rank-2, both dims dynamic) still correctly fall back to `MaxTokenLength`

### Regression tests
- Existing embedding models (all-MiniLM-L6-v2): no behavior change
- QA models (minilm-uncased-squad2): still works after Issue 2 fix (rank-2 logits still get fallback)
- Classification/NER/Reranking: no behavior change (static last dim, fallback never triggers)

---

## Acceptance Criteria

- [ ] `hasPooledOutput` correctly identifies pooled outputs with dynamic batch dimension `[-1, hidden]`
- [ ] `MaxTokenLength` fallback only applies to rank-2 non-pooled outputs (QA logits)
- [ ] Pooled and rank-3 outputs with dynamic hidden dim throw clear errors instead of silently using wrong values
- [ ] Helper method reduces code duplication in `DiscoverModelMetadata()`
- [ ] All existing model types (embedding, classification, NER, reranking, QA) continue to work

### PR DESCRIPTION
**Pull request by luisquintanilla**
## Summary

Three improvements to DiscoverModelMetadata() in OnnxTextModelScorerEstimator:

### 1. Fix hasPooledOutput detection
Change from `!dims.Contains(-1) && dims.Length == 2` to `dims.Length == 2 && dims.Last() > 0`. A pooled output like sentence_embedding with shape [-1, 384] (dynamic batch, static hidden) was incorrectly classified as non-pooled because the batch dim is -1. The batch dimension being dynamic is normal and irrelevant to pooling status.

### 2. Scope MaxTokenLength fallback
Only allow fallback to MaxTokenLength for rank-2 non-pooled outputs (QA logits where last dim = sequence length). For pooled and rank-3 outputs where last dim = hidden/embedding dimension, throw InvalidOperationException instead of silently using a wrong value.

### 3. Extract ResolveLastDimension() helper
Reduces the repeated 'get last dim, check if dynamic, fallback or throw' pattern across 5 branches into a single method with an allowDynamicFallback flag.

## Changes

**OnnxTextModelScorerEstimator.cs** - 1 file, +28 -18

## Impact

- **Latent bug fix**: No known models trigger today, but prevents incorrect behavior with exotic ONNX exports that have dynamic hidden dimensions on pooled outputs
- **Existing models**: No behavior change (static dims, fallback never triggers)

Fixes #28

### CODE REVIEW

File: src/MLNet.TextInference.Onnx/OnnxTextModelScorerEstimator.cs [hunk 1]
```diff
@@ -184,10 +184,8 @@ internal OnnxModelMetadata DiscoverModelMetadata(InferenceSession session)
         {
             outputName = _options.OutputTensorName;
             var dims = outputMeta[outputName].Dimensions;
-            hasPooledOutput = !dims.Contains(-1) && dims.Length == 2;
-            hiddenDim = (int)dims.Last();
-            if (hiddenDim <= 0)
-                hiddenDim = _options.MaxTokenLength;
+            hasPooledOutput = dims.Length == 2 && (int)dims.Last() > 0;
+            hiddenDim = ResolveLastDimension(outputName, (int)dims.Last(), allowDynamicFallback: true);
             outputRank = dims.Length;
```
[the remark this answers is not in the document]
Comment (luisquintanilla) [reply src/MLNet.TextInference.Onnx/OnnxTextModelScorerEstimator.cs:189]: **Rejected.** Let me walk through the specific concern:

> `hasPooledOutput=true` for rank-2 logits outputs like QA `start_logits` (`[-1, seqLen]`)

In practice, QA logit outputs always have **both** dimensions dynamic: `[-1, -1]` (dynamic batch, dynamic sequence). So `dims.Last() = -1`, and `hasPooledOutput = dims.Length == 2 && (int)dims.Last() > 0 = false`. The concern doesn't apply.

For the hypothetical case of `[-1, 128]` (static sequence length)  I checked the downstream consumers. `HasPooledOutput` is used by the **embedding** pipeline (`OnnxTextEmbeddingEstimator`, lines 71-72) to set `IsPrePooled` and `SequenceLength`. The **QA** pipeline uses `PreferredOutputNames = ['start_logits']` (not `OutputTensorName`) and **never reads** `HasPooledOutput`.

So even in the hypothetical, the only way this could cause harm is if a user explicitly set `OutputTensorName = 'start_logits'` (overriding auto-discovery) **and** used the embedding facade (wrong facade for QA) **and** the model had static sequence length (unusual). That's three compounding misuses.

The old check (`!dims.Contains(-1)`) was provably wrong for pooled outputs with dynamic batch  the actual bug #28 was filed to fix. The new check correctly identifies pooled outputs at the cost of a theoretical edge case with no practical impact.

File: src/MLNet.TextInference.Onnx/OnnxTextModelScorerEstimator.cs [hunk 2]
```diff
@@ -197,20 +195,16 @@ internal OnnxModelMetadata DiscoverModelMetadata(InferenceSession session)
             {
                 outputName = preferred;
                 var dims = outputMeta[preferred].Dimensions;
-                hasPooledOutput = !dims.Contains(-1) && dims.Length == 2;
-                hiddenDim = (int)dims.Last();
-                if (hiddenDim <= 0)
-                    hiddenDim = _options.MaxTokenLength;
+                hasPooledOutput = dims.Length == 2 && (int)dims.Last() > 0;
+                hiddenDim = ResolveLastDimension(preferred, (int)dims.Last(), allowDynamicFallback: true);
                 outputRank = dims.Length;
```
[the remark this answers is not in the document]
Comment (luisquintanilla) [reply src/MLNet.TextInference.Onnx/OnnxTextModelScorerEstimator.cs:200]: **Rejected.** The `PreferredOutputNames` branch is intentionally kept with `allowDynamicFallback: true` per the issue #28 spec. This branch handles **user-directed** output selection (task facades pass their hints here), not auto-discovery. The QA facade passes `PreferredOutputNames = ['start_logits']` which has dynamic last dim `[-1]`, making the fallback both correct and necessary.

For a `PreferredOutputNames` match to have a non-logit rank-3 output with dynamic hidden dim, you'd need a model where `last_hidden_state` has dynamic hidden dimension  which doesn't exist in practice (hidden dim is always a static architecture constant like 384, 768, etc.).

Issue #28 specifically scoped the throw-instead-of-fallback change to the **auto-discovery** branches only (pooled at line ~224, rank-3 unpooled at line ~236), not the user-specified branches.

File: src/MLNet.TextInference.Onnx/OnnxTextModelScorerEstimator.cs [hunk 3]
```diff
@@ -197,20 +195,16 @@ internal OnnxModelMetadata DiscoverModelMetadata(InferenceSession session)
...
             {
                 outputName = outputMeta.Keys.First();
                 var dims = outputMeta[outputName].Dimensions;
-                hasPooledOutput = !dims.Contains(-1) && dims.Length == 2;
-                hiddenDim = (int)dims.Last();
-                if (hiddenDim <= 0)
-                    hiddenDim = _options.MaxTokenLength;
+                hasPooledOutput = dims.Length == 2 && (int)dims.Last() > 0;
+                hiddenDim = ResolveLastDimension(outputName, (int)dims.Last(), allowDynamicFallback: true);
                 outputRank = dims.Length;
```
[the remark this answers is not in the document]
Comment (luisquintanilla) [reply src/MLNet.TextInference.Onnx/OnnxTextModelScorerEstimator.cs:208]: **Rejected.** Same reasoning as the comment above  this is the fallback-to-first-output sub-branch of `PreferredOutputNames`, which is still user-directed (the user set `PreferredOutputNames` but none matched). The dynamic fallback is intentionally preserved here per the issue #28 scope.
